dialogue-core/src/main/java/com/palantir/dialogue/core/StickyEndpointChannelCache.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * (c) Copyright 2025 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.EndpointChannel;
+import com.palantir.dialogue.HttpMethod;
+import java.util.function.Function;
+
+public class StickyEndpointChannelCache {
+    record StickyCacheKey(String serviceName, HttpMethod httpMethod, String endpointName) {}
+
+    private Cache<StickyCacheKey, EndpointChannel> cache =
+            Caffeine.newBuilder().maximumSize(1000).weakValues().build();
+
+    public EndpointChannel getChannel(Endpoint endpoint, Function<Endpoint, EndpointChannel> channelFactory) {
+        return cache.get(
+                new StickyCacheKey(endpoint.serviceName(), endpoint.httpMethod(), endpoint.endpointName()),
+                _unused -> channelFactory.apply(endpoint));
+    }
```

Comment:
We should probably make this a loading cache, so we guarantee the `channelFactory` is always the same (it is, but that's not enforced by this API).

Reply:
I'll address this comment + https://github.com/palantir/dialogue/pull/2647#discussion_r2261259251 here.

`channelFactory` it's potentially different. The issue is that the lifecycle of the cache is associated with the factory that can construct any kind of client (i.e. the same class constructs `FooBlocking.class` or `BarBlocking.class`). And the `channelFactory` that we're getting here is an endpointChannel factory (not the outer channel factory), which is constructed per `DialogueChannel` - so we don't have it at the time of cache creation.

Note that, for a given `endpoint` - it's likely that the `channelFactory` will have the same configs, so this is not an issue in practice. But it's hard to make the code apply this restriction.